src/spikeinterface/sortingcomponents/peak_detection.py

```diff
@@ -105,6 +109,224 @@ def detect_peaks(recording, method='by_channel', pipeline_nodes=None,
     return outs
 
 
+expanded_base_peak_dtype = np.dtype(base_peak_dtype + [("iteration", "int8")])
+
+
+class IterativePeakDetector(PeakDetector):
+    """
+    A class that iteratively detects peaks in the recording by applying a peak detector, waveform extraction,
+    and waveform denoising node. The algorithm runs for a specified number of iterations or until no peaks are found.
+    """
+
+    def __init__(
+        self,
+        recording: BaseRecording,
+        peak_detector_node: PeakDetector,
+        waveform_extraction_node: WaveformsNode,
+        waveform_denoising_node,
+        num_iterations: int = 2,
+        return_output: bool = True,
+    ):
+        """
+        Initialize the iterative peak detector.
+
+        Parameters
+        ----------
+        recording : BaseRecording
+            The recording to process.
+        peak_detector_node : PeakDetector
+            The peak detector node to use.
+        waveform_extraction_node : WaveformsNode
+            The waveform extraction node to use.
+        waveform_denoising_node
+            The waveform denoising node to use.
+        num_iterations : int, optional, default=2
+            The number of iterations to run the algorithm.
+        return_output : bool, optional, default=True
+        """
+        PeakDetector.__init__(self, recording, return_output=return_output)
+        self.peak_detector_node = peak_detector_node
+        self.waveform_extraction_node = waveform_extraction_node
+        self.waveform_denoising_node = waveform_denoising_node
+        self.num_iterations = num_iterations
+
+    def get_trace_margin(self) -> int:
+        """
+        Calculate the maximum trace margin from the internal pipeline.
+        Using the strategy as use by the Node pipeline
+
+
+        Returns
+        -------
+        int
+            The maximum trace margin.
+        """
+        internal_pipeline = (self.peak_detector_node, self.waveform_extraction_node, self.waveform_denoising_node)
+        pipeline_margin = (node.get_trace_margin() for node in internal_pipeline if hasattr(node, "get_trace_margin"))
+        return max(pipeline_margin)
+
+    def compute(self, traces_chunk, start_frame, end_frame, segment_index, max_margin) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Perform the iterative peak detection, waveform extraction, and denoising.
+
+        Parameters
+        ----------
+        traces_chunk : array-like
+            The chunk of traces to process.
+        start_frame : int
+            The starting frame for the chunk.
+        end_frame : int
+            The ending frame for the chunk.
+        segment_index : int
+            The segment index.
+        max_margin : int
+            The maximum margin for the traces.
+
+        Returns
+        -------
+        tuple of ndarray
+            A tuple containing a single ndarray with the detected peaks.
+        """
+
+        traces_chunk = traces_chunk.astype("float32")
```

Comment:
To avoid creating a new array at each iteration I think we could make a copy of this.
And this make a substraction inplace in the loop. No ?